jalantechnologies/rflask-boilerplate
# Feat: Add support for OTP based Login/SignUp

## Description
_This PR adds the support of Login/SignUp using phone number._

**NOTE**: You can use this `+91 9999999999` as phone_number and `1111` as otp for login/signup using preview link.

## Screenshot
| ![Screenshot 2024-09-04 at 10 28 14](https://github.com/user-attachments/assets/10667202-4256-44fa-83d9-fa3384b77d98) | ![Screenshot 2024-09-04 at 10 28 49](https://github.com/user-attachments/assets/86bba975-d654-48fb-ae80-6eaa04d5422b) |
| ---- | ---- |
| ![Screenshot 2024-09-04 at 10 29 02](https://github.com/user-attachments/assets/e8febf09-6b84-4d98-a442-d1c25e342c69) | ![Screenshot 2024-09-04 at 10 29 15](https://github.com/user-attachments/assets/be23f610-3b0d-444a-8c89-30cc2ab8b6d1) |


# Video
- _Login via Default Phone Number and OTP._

https://github.com/user-attachments/assets/cea0e3db-25de-46bf-9199-a71905d41801

- _Login via any number and OTP from console in development environment._


https://github.com/user-attachments/assets/eed2b776-41a2-4c01-a354-dededb6c9006



## Database schema changes
- Added an attribute in account schema
```
account {
...
phone_number: {
  country_code: str
  phone_number: str
  }
}
```
- Added a new otp schma
```
otp {
    id: ObjectId
    active: bool
    otp_code: str
    phone_number: {
        country_code: str
        phone_number: str
    }
    status: str
}
```

## Tests
### Automated test cases added
- Access Token 
  - API POST /access-tokens
    - should provide access token with phone number and correct otp code
    - should raise `OTP_ERR_01` if wrong otp code is provided
    - should raise `ACCOUNT_ERR_02` if wrong phone_number is provided
    - should raise `OTP_ERR_02` if expired otp code is provided
  - Services
    - Should successfully provide access token with phone number and otp code
    - should successfully verify access token with phone number and otp code
- Account
  - API POST /account
    - should create a new account if it does not exist and send OTP when a user signs up through the phone number
    - should send OTP for the already existing account when a user signs up through the phone number
    - should throw an error `OTP_ERR_03` when creating a new account with an invalid phone number
  - Services
    - should create new account with phone number provided
    - should throw `ACCOUNT_ERR_02` when fetching a non-existing user with provided phone number

### Manual test cases run
_For each manual test case, list the steps to test or reproduce the PR._
- Able to send OTP on the given phone number
- Able to verify OTP for the given phone number
- Error is displayed on the screen when entering the invalid phone number
- Getting error for incorrect OTP
- Able to resend OTP via Resend Button
- Successfully able to login into the dashboard with the correct OTP
